solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java

```diff
@@ -910,8 +910,12 @@ protected void createMainQuery(ResponseBuilder rb) {
         additionalAdded = addFL(additionalFL, "score", additionalAdded);
       }
     } else {
-      // reset so that only unique key is requested in shard requests
-      sreq.params.set(CommonParams.FL, rb.req.getSchema().getUniqueKeyField().getName());
+      if (rb.req.getSearcher().enableLazyFieldLoading) {
+        // reset so that only unique key is requested in shard requests
+        sreq.params.set(CommonParams.FL, rb.req.getSchema().getUniqueKeyField().getName());
+      } else {
+        sreq.params.set(CommonParams.FL, "*");
```

Comment:
@shalinmangar If lazy field loading isn't enabled, I don't think this actually changes the behavior of `SolrIndexSearcher`, since it was previously ignoring the fields list anyway. What it *should* do is allow certain distributed queries, like the ones in `DistribJoinFromCollectionTest`, to co-exist with `SolrIndexSearcher#doc()` respecting the `fields` set instead of just discarding it.